src/lib.rs

```diff
@@ -1,506 +1,632 @@
+        write!(fmt, "MaxLogLevelFilter")
     }
-    level <= DEFAULT_LOG_LEVEL
 }
 
-/// Initialize logging for the current process.
+impl MaxLogLevelFilter {
+    /// Gets the current maximum log level filter.
+    pub fn get(&self) -> LogLevelFilter {
+        max_log_level()
+    }
+
+    /// Sets the maximum log level.
+    pub fn set(&self, level: LogLevelFilter) {
+        MAX_LOG_LEVEL_FILTER.store(level as usize, Ordering::Relaxed)
+    }
+}
+
+/// Returns the current maximum log level.
 ///
-/// This is not threadsafe at all, so initialization is performed through a
-/// `Once` primitive (and this function is called from that primitive).
-fn init() {
-    let (mut directives, filter) = match os::getenv("RUST_LOG") {
-        Some(spec) => directive::parse_logging_spec(spec.as_slice()),
-        None => (Vec::new(), None),
-    };
-
-    // Sort the provided directives by length of their name, this allows a
-    // little more efficient lookup at runtime.
-    directives.sort_by(|a, b| {
-        let alen = a.name.as_ref().map(|a| a.len()).unwrap_or(0);
-        let blen = b.name.as_ref().map(|b| b.len()).unwrap_or(0);
-        alen.cmp(&blen)
-    });
+/// The `log!`, `error!`, `warn!`, `info!`, `debug!`, and `trace!` macros check
+/// this value and discard any message logged at a higher level. The maximum
+/// log level is set by the `MaxLogLevel` token passed to loggers.
+#[inline(always)]
+pub fn max_log_level() -> LogLevelFilter {
+    unsafe { mem::transmute(MAX_LOG_LEVEL_FILTER.load(Ordering::Relaxed)) }
+}
 
-    let max_level = {
-        let max = directives.iter().max_by(|d| d.level);
-        max.map(|d| d.level).unwrap_or(DEFAULT_LOG_LEVEL)
-    };
+/// Sets the global logger.
+///
+/// The `make_logger` closure is passed a `MaxLogLevel` object, which the
+/// logger should use to keep the global maximum log level in sync with the
+/// highest log level that the logger will not ignore.
+///
+/// This function may only be called once in the lifetime of a program. Any log
+/// events that occur before the call to `set_logger` completes will be
+/// ignored.
+///
+/// This function does not typically need to be called manually. Logger
+/// implementations should provide an initialization method that calls
+/// `set_logger` internally.
+pub fn set_logger<M>(make_logger: M) -> Result<(), SetLoggerError>
+        where M: FnOnce(MaxLogLevelFilter) -> Box<Log> {
+    if LOGGER.compare_and_swap(UNINITIALIZED, INITIALIZING, Ordering::Relaxed) != UNINITIALIZED {
+        return Err(SetLoggerError);
+    }
 
-    unsafe {
-        LOG_LEVEL = max_level;
+    let logger = Box::new(make_logger(MaxLogLevelFilter(())));
+    let logger = unsafe { mem::transmute::<Box<Box<Log>>, usize>(logger) };
+    LOGGER.store(logger, Ordering::Release);
+    rt::at_exit(|| {
+        // Set to INITIALIZING to prevent re-initialization after
+        let logger = LOGGER.swap(INITIALIZING, Ordering::Acquire);
```

Comment:
I commented down below as well, but keeping more relaxed orderings in `max_log_level` is totally fine but in non-perf-sensitive locations it may be better to stick to `SeqCst` for now.